sympy/functions/elementary/piecewise.py

```diff
@@ -222,9 +222,40 @@ def _eval_simplify(self, **kwargs):
         return piecewise_simplify(self, **kwargs)
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
-        for e, c in self.args:
-            if c == True or c.subs(x, 0) == True:
-                return e.as_leading_term(x)
+        if cdir == S.Zero:
+            l = self._eval_as_leading_term(x, logx=None, cdir=-1)
+            r = self._eval_as_leading_term(x, logx=None, cdir=1)
+            if l != r:
+                raise ValueError("Two sided limit of %s around 0 does not exist" % self)
+            return r
+
+        expr_pairs = self.as_expr_set_pairs()
```

Comment:
What happens if the Piecewise has more than one symbol:
```python
In [16]: p = Piecewise((x, x>y), (y, True))

In [17]: p.as_expr_set_pairs()
---------------------------------------------------------------------------
NotImplementedError 
```

Reply:
Yes we should make a note of these things but I think these can be also be done through a subsequent PR. If you see the issues raised most issues a related to getting limits for simple piecewise functions working and I think that should be the focus. During my GSoC year (2022) ... I remember we couldn't get the limits for piecewise working because, there were so many ways to represent the exact same piecewise function 
```
    p0 = Piecewise((0, Or(x < -2, x > 2)), (1, True))
    p1 = Piecewise((0, x < -2), (0, x > 2), (1, True))
    p2 = Piecewise((0, x > 2), (0, x < -2), (1, True))
    p3 = Piecewise((0, x < -2), (1, x < 2), (0, True))
    p4 = Piecewise((0, x > 2), (1, x > -2), (0, True))
    p5 = Piecewise((1, And(-2< x, x < 2)), (0, True))
   
```
And we couldn't figure out a way to find limits let's say at 0 that would satisfy all these functions. The more cases we tried addressing the more unnecessary long and buggy the code became, hence my thought here would be to get simple equivalent piecewise functions (like the 6 above) to work together. 
limit at any point `x0` for all equivalent piecewise functions should result into the same output.

Reply:
I think the tests added by @arnabnandikgp  are solid and exactly picks up from the place where we got blocked last time. I think we should focus on having good concrete code atleast to resolve these test cases first and then think of other more complex cases @oscarbenjamin

Reply:
> Yes we should make a note of these things but I think these can be also be done through a subsequent PR.

I disagree: the code added in this PR should not add more bugs. 

Using `as_expr_set_pairs` here clearly does not work for many inputs. There should be some checking before using `as_expr_set_pairs` for cases that this code can actually be expected to handle:
```python
In [9]: p = Piecewise((x, sin(x) > 0), (-x, True))

In [10]: p
Out[10]: 
⎧x   for sin(x) > 0
⎨                  
⎩-x    otherwise   

In [11]: p.as_expr_set_pairs()
---------------------------------------------------------------------------
NotImplementedError
```

Reply:
> I think the tests added by @arnabnandikgp are solid

None of them includes more than one symbol.

Reply:
As evident `as_expr_set_pairs()` method is not well suited for multivariable expressions. Maybe I can try to implement something for more than one variables, was looking into the existing methods in piecewise file and stumbled upon  `_intervals`
but for this multivariable example we have 
```
In [32]: p = Piecewise((x, x>y), (y, True))

In [33]: p._intervals(x)
Out[33]: (True, [(y, oo, x, 0), (-oo, oo, y, 1)]) # should have been (True, [(y, oo, x, 0), (-oo, y, y, 1)])

```
I will try to look into it and fix it perhaps.
But again methods like `_eval_as_leading_term` are extensively used only in limit routines and even if we handle multivariable case here it wont be of much use as currently limits don't allow limits about symbolic points anyway,
```
In [15]: p
Out[15]: 
⎧x  for x > y
⎨            
⎩y  otherwise
In [16]: limit(p,x,x,"+")
NotImplementedError: Limits approaching a variable point are not supported (x -> x) 
In [17]: limit(p,x,y,"+")
Out[17]: Limit(Piecewise((x, x > y), (y, True)), x, y, dir='+')
In [18]: limit(p,x,oo,"+")
Out[18]: Limit(Piecewise((x, x > y), (y, True)), x, oo, dir='-')
```
and even if we seek to find limits about some other numeric point we will require more info about the relation ship between the variables.
Maybe I am missing something and you might have some other example in mind that you want to see work, 
but nevertheless I can try to work on it but I don't see much use of it currently.

Reply:
> Maybe I can try to implement something for more than one variables

An alternative is just to check for the cases that the method here can handle.

We don't need to solve everything in one PR but the code should be written to handle some set of cases for which you know it can work. It should detect when those cases arise rather than returning incorrect results or failing with an error for other cases.

Reply:
> It should detect when those cases arise rather than returning incorrect results or failing with an error for other cases.

I am not sure what can be done for multivariable cases here given the fact that among all the current methods present in the piecewise file none gives us the desired results as I pointed out with the example of `_interval` above , but for the examples @oscarbenjamin showed above, the current implementation raises suitable errors, but untill we have a suitable function to kinda flatten each branch of a piecewise in multiple variables' s condition  into ranges, ig this is what we can achieve for now, with the single variable problems.

Reply:
> I am not sure what can be done for multivariable cases

We can always just return an unevaluated Limit. That is why it is useful to be able to represent the unevaluated result. If we don't yet have a way to evaluate that is always correct then we can fall back on returning the unevaluated result.

That makes it possible to write an implementation that checks for cases that can be handled but does not have to do something incorrect for cases that it cannot (yet) handle.

If the implementation here only works for Piecewise conditions that have exactly one symbol and are all linear inequalities etc then that is what should be checked.

The fact that the existing Piecewise routines don't do exactly what is needed does not matter. Just write a new function that does do what is needed and that is known to be correct for the cases that we want to handle here.